packages/eslint-plugin/src/rules/member-ordering.ts

```diff
@@ -40,6 +86,215 @@ const allMemberTypes = ['field', 'method', 'constructor'].reduce<string[]>(
   [],
 );
 
+const neverConfig = {
+  type: 'string',
+  enum: ['never'],
+};
+
+const allMemberTypesArrayConfig = {
+  type: 'array',
+  items: {
+    enum: allMemberTypes,
+  },
+};
+
+const allMemberTypesDefaultConfig = {
+  type: 'string',
+  enum: ['never'],
+};
+
+const allMemberTypesObjectConfig = {
+  type: 'object',
+  properties: {
+    memberTypes: {
+      oneOf: [allMemberTypesDefaultConfig, allMemberTypesArrayConfig],
+    },
+    order: {
+      type: 'string',
+      enum: ['alphabetically'],
+    },
+  },
+  additionalProperties: false,
+};
+
+const functionExpressions = [
+  AST_NODE_TYPES.FunctionExpression,
+  AST_NODE_TYPES.ArrowFunctionExpression,
+];
+
+/**
+ * Gets the node type.
+ * @param node the node to be evaluated.
+ */
+function getNodeType(
+  node: TSESTree.ClassElement | TSESTree.TypeElement,
+): string | null {
+  switch (node.type) {
+    case AST_NODE_TYPES.MethodDefinition:
+      return node.kind;
+    case AST_NODE_TYPES.TSMethodSignature:
+      return 'method';
+    case AST_NODE_TYPES.TSConstructSignatureDeclaration:
+      return 'constructor';
+    case AST_NODE_TYPES.ClassProperty:
+      return node.value && functionExpressions.includes(node.value.type)
+        ? 'method'
+        : 'field';
+    case AST_NODE_TYPES.TSPropertySignature:
+      return 'field';
+    default:
+      return null;
+  }
+}
+
+/**
+ * Gets the member name based on the member type.
+ *
+ * @param node the node to be evaluated.
+ * @param sourceCode
+ */
+function getMemberName(
+  node: TSESTree.ClassElement | TSESTree.TypeElement,
+  sourceCode: TSESLint.SourceCode,
+): string | null {
+  switch (node.type) {
+    case AST_NODE_TYPES.TSPropertySignature:
+    case AST_NODE_TYPES.TSMethodSignature:
+    case AST_NODE_TYPES.ClassProperty:
+      return util.getNameFromPropertyName(node.key);
+    case AST_NODE_TYPES.MethodDefinition:
+      return node.kind === 'constructor'
+        ? 'constructor'
+        : util.getNameFromClassMember(node, sourceCode);
+    case AST_NODE_TYPES.TSConstructSignatureDeclaration:
+      return 'new';
+    default:
+      return null;
+  }
+}
+
+/**
+ * Gets the member identifier (null if there is no identifier).
+ *
+ * @param node the node to be evaluated.
+ * @param sourceCode
+ */
+function getIdentifier(
+  node: TSESTree.ClassElement | TSESTree.TypeElement,
+  sourceCode: TSESLint.SourceCode,
+): string | null {
+  switch (node.type) {
+    case AST_NODE_TYPES.TSPropertySignature:
+    case AST_NODE_TYPES.TSMethodSignature:
+    case AST_NODE_TYPES.ClassProperty:
+      return util.getNameFromPropertyName(node.key);
+    case AST_NODE_TYPES.MethodDefinition:
+      return util.getNameFromClassMember(node, sourceCode);
+    default:
```

Comment:
1)

the `ClassElement` type in ts-estree is wrong, it should't have `FunctionExpression` or `TSEmptyBodyFunctionExpression`
Feel free to correct it (packages/typescript-estree/src/ts-estree/ts-estree.ts : 247)

2)

you should probably add handling for `TSAbstractMethodDefinition` and `TSAbstractClassProperty`, as they have valid names just like normal properties / methods.


---

The valid things that I can think of:
```ts
class Foo {
  // methods
  foo() {};
  ['foo']() {}
  "foo"() {};
  
  // accessors (still methods
  get foo() {};
  get ["foo"]() {};
  get "foo"() {};
  set foo() {};
  set ["foo"]() {};
  set "foo"() {};
 
  // properties
  foo;
  ['foo'];
  "foo";
  
  // abstract
  abstract foo();
  abstract ["foo"]();
  abstract "foo"();
  abstract foo;
  abstract ["foo"];
  abstract 'foo';

  // index sig
  [a: string]: number;
}
```

Reply:
I've corrected the `ClassElement` type

Reply:
If I add `TSAbstractClassProperty`, one of the "old" tests fails. I think the test isn't correct but I'm not 100% sure. Maybe it's better if you check it. I've created a separate commit for the latest review changes.

Reply:
abstract things were never implemented in the original version of the rule.
So this failure is fine - removing it like you did is the correct action.

#1004 exists to add full support for them.

I'll make sure the two PRs merged into the same release.